internal/controller/api/v1alpha1/agentserviceconfig_types.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	conditionsv1 "github.com/openshift/custom-resource-status/conditions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AgentServiceConfigSpec defines the desired state of AgentServiceConfig
+type AgentServiceConfigSpec struct{}
```

Comment:
Did you want to add to this in this PR or wait for a follow up?

Presumably one would move the existing knobs from where they are now (subscription config?) and others would configure more as needed, right?

Reply:
Every single environment variable read in by the assisted-service _could_ be configured by the subscription config and I see **a lot**. I find it most likely that you mean just the `ISO_IMAGE_TYPE`, is that right?

How hard is it for us to define all the knobs that we would **like to** expose to end-users?

Reply:
The real issue is that we will add more and more configuration, but i think that because most of the things have smart defaults and we don't expect them to change so the configuration here should be basic as well. 
Maybe we need some way to set generic env in in case we forgot to add something to the spec.
```
grep "envconfig:" -r . | awk '{print $4}' | sort --unique | wc -l
132
```
grep "envconfig:" -r . | awk '{print $4}' | sort --unique
`envconfig:"ADMIN_USERS"
`envconfig:"AGENT_DOCKER_IMAGE"
`envconfig:"AGENT_TIMEOUT_START"
`envconfig:"ALLOWED_DOMAINS"
`envconfig:"AUTH_TYPE"
`envconfig:"AUTH_TYPE"`
`envconfig:"AWS_ACCESS_KEY_ID"
`envconfig:"AWS_ACCESS_KEY_ID"`
`envconfig:"AWS_ACCESS_KEY_ID_PUBLIC"`
`envconfig:"AWS_SECRET_ACCESS_KEY"
`envconfig:"AWS_SECRET_ACCESS_KEY"`
`envconfig:"AWS_SECRET_ACCESS_KEY_PUBLIC"`
`envconfig:"BASE_DNS_DOMAINS"
`envconfig:"CHECK_CLUSTER_VERSION"
`envconfig:"CLUSTER_MONITOR_BATCH_SIZE"
`envconfig:"CLUSTER_MONITOR_INTERVAL"
`envconfig:"CLUSTER_NETWORK_CIDR"
`envconfig:"CLUSTER_NETWORK_HOST_PREFIX"
`envconfig:"CONCURRENT_ISO_EDITS"
`envconfig:"CONTROLLER_IMAGE"
`envconfig:"CREATE_S3_BUCKET"
`envconfig:"DB_HOST"`
`envconfig:"DB_NAME"`
`envconfig:"DB_PASS"`
`envconfig:"DB_PORT"`
`envconfig:"DB_USER"`
`envconfig:"DELETED_INACTIVE_AFTER"
`envconfig:"DELETED_UNREGISTERED_AFTER"
`envconfig:"DELETION_WORKER_INTERVAL"
`envconfig:"DEPLOY_TARGET"
`envconfig:"DEREGISTER_WORKER_INTERVAL"
`envconfig:"DUMMY_IGNITION"`
`envconfig:"EC_PUBLIC_KEY_PEM"`
`envconfig:"ENABLE_AUTO_RESET"
`envconfig:"ENABLE_DELETE_UNREGISTER_GC"
`envconfig:"ENABLE_DEREGISTER_INACTIVE_GC"
`envconfig:"ENABLE_KUBE_API"
`envconfig:"ENABLE_SINGLE_NODE_DNSMASQ"
`envconfig:"HOST_LOG_COLLECTION_TIMEOUT"
`envconfig:"HOST_LOG_PENDING_TIMEOUT"
`envconfig:"HOST_MONITOR_BATCH_SIZE"
`envconfig:"HOST_MONITOR_INTERVAL"
`envconfig:"HW_VALIDATOR_MAX_TIME_DIFF_MINUTES"
`envconfig:"HW_VALIDATOR_MIN_CPU_CORES"
`envconfig:"HW_VALIDATOR_MIN_CPU_CORES_MASTER"
`envconfig:"HW_VALIDATOR_MIN_CPU_CORES_WORKER"
`envconfig:"HW_VALIDATOR_MIN_DISK_SIZE_GIB"
`envconfig:"HW_VALIDATOR_MIN_RAM_GIB"
`envconfig:"HW_VALIDATOR_MIN_RAM_GIB_MASTER"
`envconfig:"HW_VALIDATOR_MIN_RAM_GIB_WORKER"
`envconfig:"IGNITION_CONFIG_BASE_FILENAME"
`envconfig:"IMAGE_EXPIRATION_INTERVAL"
`envconfig:"IMAGE_EXPIRATION_TIME"
`envconfig:"INSTALLATION_TIMEOUT"
`envconfig:"INSTALLER_IMAGE"
`envconfig:"INSTALL_RH_CA"
`envconfig:"INVENTORY"`
`envconfig:"IPV6_SUPPORT"
`envconfig:"ISO_CACHE_DIR"
`envconfig:"ISO_IMAGE_TYPE"
`envconfig:"ISO_WORKSPACE_BASE_DIR"
`envconfig:"JOB_CPU_LIMIT"
`envconfig:"JOB_CPU_REQUESTS"
`envconfig:"JOB_MEMORY_LIMIT"
`envconfig:"JOB_MEMORY_REQUESTS"
`envconfig:"JOB_MONITOR_INTERVAL"
`envconfig:"JOB_RETRY_ATTEMPTS"
`envconfig:"JOB_RETRY_INTERVAL"
`envconfig:"JWKS_CERT"`
`envconfig:"JWKS_URL"
`envconfig:"LEADER_LEASE_DURATION"
`envconfig:"LEADER_RENEW_DEADLINE"
`envconfig:"LEADER_RETRY_INTERVAL"
`envconfig:"LOG_COLLECTION_TIMEOUT"
`envconfig:"LOG_FORMAT"
`envconfig:"LOG_LEVEL"
`envconfig:"LOG_PENDING_TIMEOUT"
`envconfig:"MAX_GC_CLUSTERS_PER_INTERVAL"
`envconfig:"NAMESPACE"
`envconfig:"NTP_DEFAULT_SERVER"`
`envconfig:"OCM_BASE_URL"
`envconfig:"OCM_HOST"`
`envconfig:"OCM_LOG_LEVEL"
`envconfig:"OCM_SELF_TOKEN"
`envconfig:"OCM_SERVICE_CLIENT_ID"
`envconfig:"OCM_SERVICE_CLIENT_SECRET"
`envconfig:"OCM_TOKEN_URL"
`envconfig:"OCS_DEPLOYMENT_TYPE"
`envconfig:"OCS_DISKS_AVAILABLE"
`envconfig:"OCS_MINIMUM_CPU_COUNT"
`envconfig:"OCS_MINIMUM_DISK"
`envconfig:"OCS_MINIMUM_HOST"
`envconfig:"OCS_MINIMUM_RAM_GB"
`envconfig:"OCS_REQUIRED_COMAPCT_MODE_CPU_COUNT"
`envconfig:"OCS_REQUIRED_COMAPCT_MODE_RAM_GB"
`envconfig:"OCS_REQUIRED_CPU_COUNT"
`envconfig:"OCS_REQUIRED_DISK_CPU_COUNT"
`envconfig:"OCS_REQUIRED_DISK_RAM_GB"
`envconfig:"OCS_REQUIRED_MASTER_WORKER_HOSTS"
`envconfig:"OCS_REQUIRED_RAM_GB"
`envconfig:"OPENSHIFT_INSTALL_RELEASE_IMAGE_MIRROR"
`envconfig:"OPENSHIFT_VERSIONS"`
`envconfig:"PREPARE_FOR_INSTALLATION_TIMEOUT"
`envconfig:"PUBLIC_CONTAINER_REGISTRIES"
`envconfig:"REGISTRY_CREDS"
`envconfig:"RELEASE_TAG"
`envconfig:"RESET_CLUSTER_TIMEOUT"
`envconfig:"S3_BUCKET"
`envconfig:"S3_BUCKET"`
`envconfig:"S3_BUCKET_PUBLIC"`
`envconfig:"S3_ENDPOINT_URL"
`envconfig:"S3_ENDPOINT_URL"`
`envconfig:"S3_ENDPOINT_URL_PUBLIC"`
`envconfig:"S3_REGION"`
`envconfig:"S3_REGION_PUBLIC"`
`envconfig:"S3_SECRET_NAME"
`envconfig:"SELF_VERSION"
`envconfig:"SERVICE_BASE_URL"`
`envconfig:"SERVICE_CA_CERT_PATH"
`envconfig:"SERVICE_IPS"
`envconfig:"SERVICE_NETWORK_CIDR"
`envconfig:"SKIP_CERT_VERIFICATION"
`envconfig:"STORAGE"
`envconfig:"SUBSYSTEM_RUN"`
`envconfig:"SUPPORT_L2"
`envconfig:"TEST_TOKEN"`
`envconfig:"TEST_TOKEN_ADMIN"`
`envconfig:"TEST_TOKEN_UNALLOWED"`
`envconfig:"WITH_AMS_SUBSCRIPTIONS"
`envconfig:"WORK_DIR"
`envconfig:"X-REQUEST-ID"

```

```

Reply:
Are we comfortable with their being no configuration for the initial implementation?

Reply:
We do want to enable configure everything but we don't want to update multiple places when adding new configuration.
So i'm suggesting to map the important configuration and leave an option to set everything in case we forget to update.